services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt

```diff
@@ -32,6 +58,264 @@ class HakemusService(
+     * yhteystieto, and it cannot have an id. If the yhteystieto is not present in the request, it
+     * means that the application yhteystieto is removed.
+     */
+    private fun assertYhteystietoValidity(
+        applicationId: Long,
+        rooli: ApplicationContactType,
+        hakemusyhteystietoEntity: HakemusyhteystietoEntity?,
+        customerWithContacts: CustomerWithContactsRequest?
+    ) {
+        if (
+            hakemusyhteystietoEntity != null &&
+                customerWithContacts != null &&
+                customerWithContacts.customer.yhteystietoId != hakemusyhteystietoEntity.id
+        ) {
+            // ids don't match
+            throw InvalidHakemusyhteystietoException(
+                applicationId,
+                rooli,
+                hakemusyhteystietoEntity.id,
+                customerWithContacts.customer.yhteystietoId
+            )
+        }
+        if (
+            hakemusyhteystietoEntity == null &&
+                customerWithContacts?.customer?.yhteystietoId != null
+        ) {
+            // new yhteystieto has an id when it shouldn't
+            throw InvalidHakemusyhteystietoException(
+                applicationId,
+                rooli,
+                null,
+                customerWithContacts.customer.yhteystietoId
+            )
+        }
+    }
+
+    /** Assert that the contacts are users of the hanke. */
+    private fun assertYhteyshenkilotValidity(
+        hanke: HankeEntity,
+        newHankekayttajaIds: Set<UUID>,
+        customMessageOnFailure: (Set<UUID>) -> String
+    ) {
+        val currentHankekayttajaIds =
+            hankeKayttajaService.getKayttajatByHankeId(hanke.id).map { it.id }.toSet()
+        val newInvalidHankekayttajaIds = newHankekayttajaIds.minus(currentHankekayttajaIds)
+        if (newInvalidHankekayttajaIds.isNotEmpty()) {
+            throw InvalidHakemusyhteyshenkiloException(
+                customMessageOnFailure(newInvalidHankekayttajaIds)
+            )
+        }
+    }
+
+    /** Assert that the geometries are compatible with the hanke area geometries. */
+    private fun assertGeometryCompatibility(
+        hankeId: Int,
+        areas: List<ApplicationArea>,
+        customMessageOnFailure: (ApplicationArea) -> String
+    ) {
+        areas.forEach { area ->
+            if (!geometriatDao.isInsideHankeAlueet(hankeId, area.geometry))
+                throw ApplicationGeometryNotInsideHankeException(customMessageOnFailure(area))
+        }
+    }
+
+    /** Update the hanke areas based on the update request work areas. */
+    private fun updateHankealueet(hankeEntity: HankeEntity, updateRequest: HakemusUpdateRequest) {
+        val hankealueet =
+            updateRequest.areas?.mapIndexed { i, area ->
+                area.toNewHankealue(i, updateRequest.startTime!!, updateRequest.endTime!!)
+            } ?: emptyList()
+        hankeEntity.alueet.clear()
+        if (hankealueet.isNotEmpty()) {
```

Comment:
Nothing will happen anyway if hankealueet is empty.

Reply:
Good call, thanks!